### PR TITLE
rabbit_peer_discovery: Rewrite the core logic

### PR DESCRIPTION
**Pull request by dumbbell**
## Why

This work started as an effort to add peer discovery support to our Khepri integration. Indeed, as part of the task to integrate Khepri, we missed the fact that `rabbit_peer_discovery:maybe_create_cluster/1` was called from the Mnesia-specific code only. Even though we knew about it because we hit many issues caused by the fact the `join_cluster` and peer discovery use different code path to create a cluster.

To add support for Khepri, the first version of this patch was to move the call to `rabbit_peer_discovery:maybe_create_cluster/1` from `rabbit_db_cluster` instead of `rabbit_mnesia`. To achieve that, it made sense to unify the code and simply call `rabbit_db_cluster:join/2` instead of duplicating the work.

Unfortunately, doing so highlighted another issue: the way the node to cluster with was selected. Indeed, it could cause situations where multiple clusters are created instead of one, without resorting to out-of-band counter-measures, like a 30-second delay added in the Kubernetes operator (rabbitmq/cluster-operator#1156). This problem was even more frequent when we tried to unify the code path and call `join_cluster`.

After several iterations on the patch and even more discussions with the team, we decided to rewrite the algorithm to make node selection more robust and still use `rabbit_db_cluster:join/2` to create the cluster.

## How

We wanted the new algorithm to fulfill the following properties:

1. `rabbit_peer_discovery` should provide the ability to re-trigger it easily to re-evaluate the cluster. The new public API is `rabbit_peer_discovery:sync_desired_cluster/0`.

2. The selection of the node to join should be designed in a way that all nodes select the same, regardless of the order in which they become available. The adopted solution is to sort the list of discovered nodes with the following criterias (in that order):

    1. the size of the cluster a discovered node is part of; sorted from bigger to smaller clusters
    2. the start time of a discovered node; sorted from older to younger nodes
    3. the name of a discovered node; sorted alphabetically

   The first node in that list will not join anyone and simply proceed with its boot process. Other nodes will try to join the first node.

3. To reduce the chance of incorrectly having multiple standalone nodes because the discovery backend returned only a single node, we want to apply the following constraints to the list of nodes after it is filtered and sorted (see property 2 above):

    * The list must contain `node()` (i.e. the node running peer discovery itself).
    * If the RabbitMQ's cluster size hint is greater than 1, the list must have at least two nodes. The cluster size hint is the maximum between the configured target cluster size hint and the number of elements in the nodes list returned by the backend.

   If one of the constraint is not met, the entire peer discovery process is restarted after a delay.

4. The lock is acquired only to protect the actual join, not the discovery step where the backend is queried to get the list of peers. With the node selection described above, this will let the first node to start without acquiring the lock.

5. The cluster membership views queried as part of the algorithm to sort the list of nodes will be used to detect additional clusters or standalone nodes that did not cluster correctly. These nodes will be asked to re-evaluate peer discovery to increase the chance of forming a single cluster.

6. After some delay, peer discovery will be re-evaluated to further eliminate the chances of having multiple clusters instead of one.

This commit covers properties from point 1 to point 4. Remaining properties will be the scope of additional pull requests after this one works.

If there is a failure at any point during discovery, filtering/sorting, locking or joining, the entire process is restarted after a delay. This is configured using the following parameters:
* `cluster_formation.discovery_retry_limit`
* `cluster_formation.discovery_retry_interval`

The default parameters were bumped to 30 retries with a delay of 1 second between each.

The locking retries/interval parameters are not used by the new algorithm anymore.

There are extra minor changes that come with the rewrite:
* The configured backend is cached in a persistent term. The goal is to make sure we use the same backend throughout the entire process and when we call `maybe_unregister/0` even if the configuration changed for whatever reason in between.
* `maybe_register/0` is called from `rabbit_db_cluster` instead of at the end of a successful peer discovery process. `rabbit_db_cluster` had to call `maybe_register/0` if the node was not virgin anyway. So make it simpler and always call it in `rabbit_db_cluster` regardless of the state of the node.
* `log_configured_backend/0` is gone. `maybe_init/0` can log the backend directly. There is no need to explicitly call another function for that.
* Messages are logged using `?LOG_*()` macros instead of the old `rabbit_log` module.

### CODE REVIEW

**Comment by dumbbell** [comment]
After several iterations on the patch and more discussions with the team, I will restart from scratch. The idea is still to use the common `join_cluster` code path instead of having something specific to peer discovery. However, this highlights some issues with the current peer discovery code. In particular, the way the node to join is selected means that e.g. node 4 joins node 1, but node 1 then joins node 2, kicking node 4 out of the cluster.

The current consensus is to have the following properties:

1. `rabbit_peer_discovery` should provide the ability to re-trigger it easily to re-evaluate the cluster.

2. The selection of the node to join should be designed in a way that all nodes select the same, regardless of the order in which they become available. The current solution is to sort the list of discovered nodes with the following criterias (in that order):
    1. the size of the cluster a discovered node is part of; sorted from bigger to smaller clusters
    2. the uptime of a discovered node; sorted from older to younger nodes
    3. the name of a discovered node; sorted alphabetically

    The first node in that list does not join anyone and simply proceed with its boot process. Other nodes try to join the first node.

3. Only the actual join will be protected by the peer discovery lock, not the discovery step where the backend is queried to get the list of peers. With the node selection described above, this will let the first node to start without acquiring the lock.

4. To reduce the chance of incorrectly having multiple standalone nodes because the discovery backend returned only a single node, we want the following failsafes:
    * If the backend returns a list that does not contain `node()`, the node will wait and retry the query.
    * If the RabbitMQ's cluster size hint is greater than 1 and the backend returns a single node, the node will wait and retry the query.
    * After a timeout, the node will abort peer discovery and boot as a standalone node.

5. The cluster membership views queried as part of the algorithm to sort the list of nodes will be used to detect standalone nodes that did not cluster correctly or additional clusters. These nodes will be asked to re-evaluate peer discovery to increase the chance of forming a single cluster.

6. After some delay, peer discovery will be re-evaluated to further eliminate the chances of having multiple clusters instead of one.

This pull request will cover properties from 1 to 3. Remaining properties will be the scope of additional pull requests after this one works.

**Comment by mkuratczyk** [comment]
Sounds great, thanks for all the work you are putting into this! Perhaps this is implicitly covered by 4B, but one things I'd add is that if the cluster size hint is `N`, a node that doesn't know see `N` nodes in the cluster should not consider itself successfully booted.

**Comment by dumbbell** [comment]
This is not covered by 4B. I clarified my comment. The idea is that if the cluster size hint is 2 or more, peer discovery should expect that the backend returns at least two nodes. This is to avoid that an early query of the backend only returns `[node()]` to every nodes.

> a node that doesn't now see N nodes in the cluster should not consider itself successfully booted.

Do you suggest that peer discovery should wait for a list of discovered nodes of size N? Or should peer discovery finish, but later in the boot process, we pause until all N nodes joined the cluster?

What about deployments where nodes are started sequentially? Or deployments where the peer discovery backend and the configured cluster hint are out-of-sync?

**Comment by michaelklishin** [comment]
If we wait for N nodes to appear, and N = the total number of nodes, a single node that fails to boot will cause issues.

Also, [Kubernetes generally assumes that nodes do not have any inter-dependencies](https://www.rabbitmq.com/cluster-formation.html#peer-discovery-k8s).

Hence the idea to re-evaluate cluster membership and retry after node boot.

**Comment by mkuratczyk** [comment]
I guess I haven't thought that through before. Indeed, with sequentially booted nodes, the nodes would need to be fully booted, for the next to even start booting (eg. that's the default behaviour for StatefulSets on Kubermetes, our Operator sets the startup policy to `Parallel`, but there are many deployments that don't use this option). I guess what I'd mostly want and I think Michael's comment covers is that ideally, the cluster should be somehow unusable until all expected nodes are there. Eg we discussed in the past the possibility of rejecting any queue declaration until all expected nodes are present - that could be a way to make it clear that something went wrong (and ideally automatically fixing/retrying cluster formation).

**Comment by dumbbell** [comment]
> If the RabbitMQ's cluster size hint is greater than 1 and the backend returns a single node, the node will wait and retry the query until at least two nodes are returned.

Small update to this: instead of using the configured target cluster size hint alone, I take the max between this value and the number of nodes returned by the backend. This is handy with the classic config backend for instance: the static list of nodes is the cluster size hint.

**Comment by dumbbell** [comment]
The branch contains two more commits which make sure that we don't connect to all discovered nodes too early and trigger a global process name registry conflict. See individual commit messages for more details.

**Review by mkuratczyk — APPROVED** [review]
We've run many tests on Kubernetes (kind and GKE) and we found no issues. It's certainly much better than current `main`, which has multiple issues.

Once merged, we can ask the community for additional testing in other environments.

File: deps/rabbit/src/rabbit_peer_discovery.erl [hunk 1]
```diff
@@ -90,142 +92,722 @@ lock_acquisition_failure_mode() ->
...
+               "Peer discovery: not satisfyied with discovered peers: the "
+               "list should contain at least two nodes with a configured "
+               "cluster size hint of ~b nodes",
+               [ClusterSizeHint],
+               #{domain => ?RMQLOG_DOMAIN_PEER_DISC})
+    end,
+
+    ThisNodeIsIncluded andalso HasEnoughNodes;
+can_use_discovered_nodes(_DiscoveredNodes, []) ->
+    ?LOG_DEBUG(
```
Comment (lukebakken) [inline deps/rabbit/src/rabbit_peer_discovery.erl:681]: This should be LOG_INFO or WARNING

**Review by lukebakken — COMMENTED** [review]
Only one more suggestion.

File: deps/rabbit/src/rabbit_mnesia.erl [hunk 2]
```diff
@@ -114,13 +114,13 @@
 init() ->
     ensure_mnesia_running(),
     ensure_mnesia_dir(),
+    %% Peer discovery may have been a no-op if it decided that all other nodes
+    %% should join this one. Therefore, we need to look at if this node is
+    %% still virgin and finish our use of Mnesia accordingly. In particular,
+    %% this second part crates all our Mnesia tables.
```
[the remark this answers is not in the document]
Comment (dumbbell) [reply deps/rabbit/src/rabbit_mnesia.erl:120]: Addressed in #10073. Thanks!

File: deps/rabbit/src/rabbit_peer_discovery.erl [hunk 3]
```diff
@@ -90,142 +92,722 @@ lock_acquisition_failure_mode() ->
...
--spec log_configured_backend() -> ok.
-
-log_configured_backend() ->
-  rabbit_log:info("Configured peer discovery backend: ~ts", [backend()]).
+-spec maybe_init() -> ok.
+%% @doc Initializes the peer discovery subsystem.
 
 maybe_init() ->
     Backend = backend(),
+    ?LOG_DEBUG(
```
[the remark this answers is not in the document]
Comment (dumbbell) [reply deps/rabbit/src/rabbit_peer_discovery.erl:100]: Addressed in #10073. Thanks!

File: deps/rabbit/src/rabbit_peer_discovery.erl [hunk 4]
```diff
@@ -90,142 +92,722 @@ lock_acquisition_failure_mode() ->
...
+    case peer:start(#{name => PeerName, args => VMArgs1}) of
+        {ok, Pid, Peer} ->
+            ?LOG_DEBUG(
+               "Peer discovery: use temporary hidden node '~ts' to query "
+               "discovered peers properties",
+               [Peer],
+               #{domain => ?RMQLOG_DOMAIN_PEER_DISC}),
+            Ret = erpc:call(Peer, ?MODULE, do_query_node_props, [Nodes]),
+            peer:stop(Pid),
+            Ret;
```
[the remark this answers is not in the document]
Comment (dumbbell) [reply deps/rabbit/src/rabbit_peer_discovery.erl:392]: Addressed in #10073. Thanks!

File: deps/rabbit/src/rabbit_peer_discovery.erl (hunk 1 above)
Comment (dumbbell) [reply deps/rabbit/src/rabbit_peer_discovery.erl:681]: Addressed in #10073. Thanks!

**Pull request merged**